docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/launcher/DockerComputerJNLPLauncher.java

```diff
@@ -0,0 +1,153 @@
+import com.github.dockerjava.api.NotFoundException;
+import com.github.dockerjava.api.command.CreateContainerCmd;
+import com.github.dockerjava.api.command.ExecCreateCmdResponse;
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.nirima.jenkins.plugins.docker.DockerComputer;
+import com.nirima.jenkins.plugins.docker.DockerTemplate;
+import com.nirima.jenkins.plugins.docker.DockerTemplateBase;
+import hudson.model.Descriptor;
+import hudson.model.TaskListener;
+import hudson.slaves.ComputerLauncher;
+import hudson.slaves.JNLPLauncher;
+import hudson.slaves.SlaveComputer;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.Charsets;
+import org.apache.commons.io.IOUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.util.Arrays;
+
+/**
+ * JNLP launcher. Doesn't require open ports on docker host.
+ *
+ * Steps:
+ * - runs container with nop command
+ * - as launch action executes jnlp connection to master
+ *
+ * @author Kanstantsin Shautsou
+ */
+public class DockerComputerJNLPLauncher extends DockerComputerLauncher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DockerComputerJNLPLauncher.class);
+
+    /**
+     * Configured from UI
+     */
+    protected JNLPLauncher jnlpLauncher;
+
+    @DataBoundConstructor
+    public DockerComputerJNLPLauncher(JNLPLauncher jnlpLauncher) {
+        this.jnlpLauncher = jnlpLauncher;
+    }
+
+    public JNLPLauncher getJnlpLauncher() {
+        return jnlpLauncher;
+    }
+
+    @Override
+    public boolean isLaunchSupported() {
+        return true;
+    }
+
+    @Override
+    public void launch(SlaveComputer computer, TaskListener listener) throws IOException, InterruptedException {
+        final PrintStream logger = listener.getLogger();
+        final DockerComputer dockerComputer = (DockerComputer) computer;
+        final String containerId = dockerComputer.getContainerId();
+        final String rootUrl = Jenkins.getInstance().getRootUrl();
+        final DockerClient connect = dockerComputer.getCloud().connect();
+
+        // exec jnlp connection in running container
+        // TODO implement PID 1 replacement
+
+        final InputStream resourceAsStream = DockerComputerJNLPLauncher.class.getResourceAsStream("init.sh");
+
+        String cdCmd = "cd /home/jenkins/";
+        String wgetSlaveCmd = "wget " + rootUrl + "jnlpJars/slave.jar -O slave.jar";
+        String jnlpConnectCmd = "java -jar slave.jar "
+                + "-jnlpUrl " + rootUrl + dockerComputer.getUrl() + "slave-agent.jnlp ";
+//                + "-secret " + dockerComputer.getJnlpMac();
+
+        String[] connectCmd = {
+                "bash", "-c", cdCmd + " && " + wgetSlaveCmd + " && " + jnlpConnectCmd
```

Comment:
@KostyaSha my preference would be to declare an `ENTRYPOINT` in my dockerized slave and have the plugin send in the `-jnlpUrl` and `-secret` args.  This would let we do things like baking the slave.jar into my docker build if I want to.